# variables (or python variables)
> -> container to store data  

variable types: (normally of these two types throughout most of the programming languages):  
**1. statistically typed**  
**2. dynamically typed**

for example  
> *C language is statistically typed*
```c
    int x = 5
```

> *while python is dynamically typed*
```python
    x = 5
```

In [1]:
# creating a variable in python
name = "@jkmloom"
code_year = 2026
weight_kg = 88.75

python considers evey variable as an object. this object has got an address and we can find this adddress using an in-built function/method -> `id`

In [4]:
# creating an variable
age = 23
# finding the object's address
print(id(age))

110994320776424


> the object contains the following info:  

**considering object id: 110994320776424**  
type            -> integer  
value           -> 23  
reference count -> 1

In [6]:
# if we've got 2 variables pointing to the same value then the variables will be sharing the same object id
x = 5
print(id(x))
y = 5
print(id(y))

110994320775848
110994320775848


> if multiple variables are pointing to the same value then the reference count changes accordengly:  

for example  **considering object id: 110994320775848** >>  
type -> integer  
value -> 5  
reference count -> 2  

---


> **rules for naming a variable** *(in python)*

1. allowed characters -->>  
    i.  a/A - z/Z (alphabets)  
    ii. _ (underscore/s)  
    iii. 0, 1, 2, 3, .. (numerical values)  
[!, @, #, $, %, ^, etc. are not allowed]  

2. first character of the variable must be a alphabet of an underscore  
[we can't use numerical values like 0, 1, 2, 3, ..]  

3. variable names are case sensative (a and A are different)  

4. avoid using keywords as variable names (not allowed)  
**keywords** -->> reserved words whose meanings are already assigned to the computer.  

5. variable name should be a single word.

In [9]:
# for example
variable_1 = "this is an example of a valid variable name in python"
_2variable = "this is another example of a valid variable name in python"

a = 23
A = 24
print(f"a = {a} and A = {A}")

a = 23 and A = 24


In [14]:
# looking for all the keywords in python
help()

Welcome to Python 3.11's help utility! If this is your first time using
Python, you should definitely check out the tutorial at
https://docs.python.org/3.11/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To get a list of available
modules, keywords, symbols, or topics, enter "modules", "keywords",
"symbols", or "topics".

Each module also comes with a one-line summary of what it does; to list
the modules whose name or summary contain a given string such as "spam",
enter "modules spam".

To quit this help utility and return to the interpreter,
enter "q" or "quit".


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as         

**for example -->>**  
```python
    def = 'invalid variable name'
```
we can't use keywords as variable name so, the above example is invalid -> it'll through an error
while the below example is an valid example:  
```python
    def1 = 'valid variable name'
```

---

> not a rule but a recommendation -->> variable name should make some sense, it should be logical.  

for example -->> storing the age of an young adult  
```python
    i = 23
```
the above example is technically correct and will nto through any error, but the name 'i' doesn't makes much sense.  
so we should name it something sensable like in the example below:  
```python
    age = 23
```